File: app/src/main/java/com/sofps/inspirationalquotes/ui/QuotesSlidePresenter.java
package com.sofps.inspirationalquotes.ui;

import com.sofps.inspirationalquotes.asynctask.QuotesLoader;
import com.sofps.inspirationalquotes.data.Quote;
import com.sofps.inspirationalquotes.data.source.QuotesRepository;
import com.sofps.inspirationalquotes.util.LanguagePreferences;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

public class QuotesSlidePresenter
        implements QuotesSlideContract.Presenter, QuotesLoader.QuotesLoaderTaskListener {

    private final QuotesSlideContract.View mView;
    private final QuotesRepository mQuotesRepository;
    private final LanguagePreferences mLanguagePreferences;

    private ArrayList<Quote> mQuotes = new ArrayList<>();

    QuotesSlidePresenter(QuotesSlideContract.View view, QuotesRepository quotesRepository, LanguagePreferences languagePreferences) {
        mView = view;
        mQuotesRepository = quotesRepository;
        mLanguagePreferences = languagePreferences;

        mView.setPresenter(this);
    }

    @Override
    public void start() {
        loadQuotesForCurrentLanguage();
    }

    private void loadQuotesForCurrentLanguage() {
        mView.showProgress(true);
        mQuotesRepository.loadQuotesForLanguage(mLanguagePreferences.getLanguage(), this);
    }

    @Override
    public void onQuotesLoaderTaskComplete(@NotNull List<Quote> quoteList) {
        mQuotes.clear();
        mQuotes.addAll(quoteList);

        mView.initialize();
        mView.setQuotes(mQuotes);
        mView.showProgress(false);
    }

    @Override
    public void onQuotesLoaderTaskInProgress() {
        mView.showProgress(true);
    }

    @Override
    public void onLanguageChange() {
        loadQuotesForCurrentLanguage();
    }

    @Override
    public void setQuotes(List<Quote> quotes) {
        mQuotes.addAll(quotes);
    }

    @Override
    public List<Quote> getQuotes() {
        return mQuotes;
    }

    @Override
    public void onQuoteShow(Quote quote) {
        mQuotesRepository.addOneTimeShowed(quote);
    }
}

File: app/src/main/java/com/sofps/inspirationalquotes/ui/QuotesSlidePresenter.kt
package com.sofps.inspirationalquotes.ui

import com.sofps.inspirationalquotes.data.source.QuotesRepository
import com.sofps.inspirationalquotes.model.QuoteModel
import com.sofps.inspirationalquotes.model.ViewState
import com.sofps.inspirationalquotes.util.LanguagePreferences
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch

class QuotesSlidePresenter internal constructor(
        private val view: QuotesSlideContract.View,
        private val quotesRepository: QuotesRepository,
        private val languagePreferences: LanguagePreferences
) : QuotesSlideContract.Presenter {

    private val internalQuotes = mutableListOf<QuoteModel>()

    override var quotes: List<QuoteModel>
        get() = internalQuotes
        set(quotes) {
            internalQuotes.addAll(quotes)
        }

    override fun start() {
        loadQuotesForCurrentLanguage()
    }

    private fun loadQuotesForCurrentLanguage() {
        GlobalScope.launch(Dispatchers.Main) {
            quotesRepository.getQuotesForLanguage(languagePreferences.language)
                    .collect { state ->
                        when (state) {
                            is ViewState.Success -> {
                                internalQuotes.clear()
                                internalQuotes.addAll(state.data)

                                view.initialize()
                                view.setQuotes(internalQuotes)
                                view.showProgress(false)
                            }
                            is ViewState.Loading -> view.showProgress(true)
                            is ViewState.Error -> {
                                // TODO show error toast("Something went wrong ¯\\_(ツ)_/¯ => ${state.message}")
                            }
                        }
                    }
        }
    }

    override fun onLanguageChange() {
        loadQuotesForCurrentLanguage()
    }

    override fun onQuoteShow(quote: QuoteModel) {
        quotesRepository.addOneTimeShowed(quote)
    }
}
